turbo-txpool/src/tx.rs

```diff
@@ -2,20 +2,95 @@ use crate::error::{Error, RlpResultExt};
 
 use ethereum_types::{Address, H256, U256};
 
+use k256::ecdsa::{self, recoverable};
+use k256::EncodedPoint;
+
+use std::convert::{TryFrom, TryInto};
+
 use tiny_keccak::{Hasher, Keccak};
 
-fn keccak(input: &[u8]) -> H256 {
+fn keccak(input: &[u8]) -> [u8; 32] {
     let mut keccak = Keccak::v256();
     keccak.update(input);
 
     let mut output = [0u8; 32];
     keccak.finalize(&mut output);
-    H256::from(output)
+    output
+}
+
+#[derive(Debug, Clone)]
+pub struct VerifiedTx {
+    hash: H256,
+    from: Address,
+    tx: Tx,
+}
+
+impl VerifiedTx {
+    pub fn new(tx: Tx) -> Result<Self, ecdsa::Error> {
+        let v = 1 - (tx.v % 2);
+        let mut r = [0u8; 32];
+        tx.r.to_big_endian(&mut r);
+        let mut s = [0u8; 32];
+        tx.s.to_big_endian(&mut s);
+
+        let sig = recoverable::Signature::new(
+            &ecdsa::Signature::from_scalars(r, s)?,
+            recoverable::Id::new(v.try_into().unwrap())?,
+        )?;
+
+        let mut stream = rlp::RlpStream::new();
+        tx.signature_encode(&mut stream);
+        let vkey = sig.recover_verify_key(stream.as_raw())?;
+
+        let point = EncodedPoint::from(&vkey).decompress();
+
+        let pubkey = match point {
+            Some(k) => k.to_bytes(),
+            None => return Err(ecdsa::Error::new()),
+        };
+
+        let from = Address::from_slice(&keccak(&pubkey[1..])[12..]);
```

Comment:
Not quite clear to me why you need `[1..]` of `pubkey`.

Reply:
That's just how it works I guess? See the [`ethers-rs` implementation](https://github.com/gakonst/ethers-rs/blob/ac93c0bc65fb31a6d8467a74e0b16a21fced3c88/ethers-core/src/types/signature.rs#L111)